libraries/APM_Control/AP_YawController.cpp

```diff
@@ -25,160 +25,351 @@ extern const AP_HAL::HAL& hal;
     float intLimScaled = _imax * 0.01f / (_K_D * scaler * scaler);
 
     // Constrain the integrator state
     _integrator = constrain_float(_integrator, -intLimScaled, intLimScaled);
-	
-	// Protect against increases to _K_D during in-flight tuning from creating large control transients
-	// due to stored integrator values
-	if (_K_D > _K_D_last && _K_D > 0) {
-	    _integrator = _K_D_last/_K_D * _integrator;
-	}
-	_K_D_last = _K_D;
-	
-	// Calculate demanded rudder deflection, +Ve deflection yaws nose right
-	// Save to last value before application of limiter so that integrator limiting
-	// can detect exceedance next frame
-	// Scale using inverse dynamic pressure (1/V^2)
-	_pid_info.I = _K_D * _integrator * scaler * scaler;
-	_pid_info.D = _K_D * (-rate_hp_out) * scaler * scaler;
-	_last_out =  _pid_info.I + _pid_info.D;
-
-	// Convert to centi-degrees and constrain
-	return constrain_float(_last_out * 100, -4500, 4500);
+
+    // Protect against increases to _K_D during in-flight tuning from creating large control transients
+    // due to stored integrator values
+    if (_K_D > _K_D_last && _K_D > 0) {
+        _integrator = _K_D_last/_K_D * _integrator;
+    }
+    _K_D_last = _K_D;
+
+    // Calculate demanded rudder deflection, +Ve deflection yaws nose right
+    // Save to last value before application of limiter so that integrator limiting
+    // can detect exceedance next frame
+    // Scale using inverse dynamic pressure (1/V^2)
+    _pid_info.I = _K_D * _integrator * scaler * scaler;
+    _pid_info.D = _K_D * (-rate_hp_out) * scaler * scaler;
+    _last_out =  _pid_info.I + _pid_info.D;
+
+    // Convert to centi-degrees and constrain
+    return constrain_float(_last_out * 100, -4500, 4500);
+}
+
+// get actuator output for direct rate control
+// desired_rate is in deg/sec. scaler is the surface speed scaler
+float AP_YawController::get_rate_out(float desired_rate, float scaler, bool disable_integrator)
+{
+    const AP_AHRS &_ahrs = AP::ahrs();
+
+    const float dt = AP::scheduler().get_loop_period_s();
+    const float eas2tas = _ahrs.get_EAS2TAS();
+    bool limit_I = fabsf(_last_out) >= 45;
+    float rate_z = _ahrs.get_gyro().z;
+    float aspeed;
+    float old_I = rate_pid.get_i();
+
+    rate_pid.set_dt(dt);
+
+    if (!_ahrs.airspeed_estimate(aspeed)) {
+        aspeed = 0;
+    }
+    bool underspeed = aspeed <= float(aparm.airspeed_min);
+    if (underspeed) {
+        limit_I = true;
+    }
+
+    // the P and I elements are scaled by sq(scaler). To use an
+    // unmodified AC_PID object we scale the inputs and calculate FF separately
+    //
+    // note that we run AC_PID in radians so that the normal scaling
+    // range for IMAX in AC_PID applies (usually an IMAX value less than 1.0)
+    rate_pid.update_all(radians(desired_rate) * scaler * scaler, rate_z * scaler * scaler, limit_I);
+
+    if (underspeed) {
+        // when underspeed we lock the integrator
+        rate_pid.set_integrator(old_I);
+    }
```

Comment:
If we have limed I there should be no need to also reset I like this.

Reply:
for hand launch we really don't want the integrator before throw off, esp as the I limit could be quite large